src/tools/nvmf-rbd-gw/nvmf-rbd-gw.c

```diff
@@ -0,0 +1,964 @@
+	 */
+	spdk_thread_lib_init_ext(nvmf_reactor_thread_op, nvmf_reactor_thread_op_supported,
+				 sizeof(struct nvmf_lw_thread));
+
+	/* Spawn one system thread per CPU core. The system thread is called a reactor.
+	 * SPDK will spawn lightweight threads that must be mapped to reactors in
+	 * nvmf_schedule_spdk_thread. Using a single system thread per CPU core is a
+	 * choice unique to this application. SPDK itself does not require this specific
+	 * threading model. For example, another viable threading model would be
+	 * dynamically scheduling the lightweight threads onto a thread pool using a
+	 * work queue.
+	 */
+	SPDK_ENV_FOREACH_CORE(i) {
+		nvmf_reactor = calloc(1, sizeof(struct nvmf_reactor));
+		if (!nvmf_reactor) {
+			fprintf(stderr, "failed to alloc nvmf reactor\n");
+			rc = -ENOMEM;
+			goto err_exit;
+		}
+
+		nvmf_reactor->core = i;
+
+		nvmf_reactor->threads = spdk_ring_create(SPDK_RING_TYPE_MP_SC, 1024, SPDK_ENV_SOCKET_ID_ANY);
+		if (!nvmf_reactor->threads) {
+			fprintf(stderr, "failed to alloc ring\n");
+			free(nvmf_reactor);
+			rc = -ENOMEM;
+			goto err_exit;
+		}
+
+		TAILQ_INSERT_TAIL(&g_reactors, nvmf_reactor, link);
+
+		if (i == master_core) {
+			g_master_reactor = nvmf_reactor;
+			g_next_reactor = g_master_reactor;
+		} else {
+			rc = spdk_env_thread_launch_pinned(i,
+							   nvmf_reactor_run,
+							   nvmf_reactor);
+			if (rc) {
+				fprintf(stderr, "failed to pin reactor launch\n");
+				goto err_exit;
+			}
+		}
+	}
+
+	/* Spawn a lightweight thread only on the current core to manage this application. */
+	spdk_cpuset_zero(&cpumask);
+	spdk_cpuset_set_cpu(&cpumask, master_core, true);
+	snprintf(thread_name, sizeof(thread_name), "nvmf_master_thread");
+	g_init_thread = spdk_thread_create(thread_name, &cpumask);
+	if (!g_init_thread) {
+		fprintf(stderr, "failed to create spdk thread\n");
+		return -1;
+	}
+
+	fprintf(stdout, "nvmf threads initlize successfully\n");
+	return 0;
+
+err_exit:
+	return rc;
+}
+
+static void
+nvmf_destroy_threads(void)
+{
+	struct nvmf_reactor *nvmf_reactor, *tmp;
+
+	TAILQ_FOREACH_SAFE(nvmf_reactor, &g_reactors, link, tmp) {
+		spdk_ring_free(nvmf_reactor->threads);
+		free(nvmf_reactor);
+	}
+
+	pthread_mutex_destroy(&g_mutex);
+	spdk_thread_lib_fini();
+	fprintf(stdout, "nvmf threads destroy successfully\n");
+}
+
+static void
+nvmf_tgt_destroy_done(void *ctx, int status)
+{
+	fprintf(stdout, "destroyed the nvmf target service\n");
+
+	g_target_state = NVMF_FINI_SUBSYSTEM;
+	nvmf_target_advance_state();
+}
+
+static void
+nvmf_destroy_nvmf_tgt(void)
+{
+	if (g_nvmf_tgt.tgt) {
+		spdk_nvmf_tgt_destroy(g_nvmf_tgt.tgt, nvmf_tgt_destroy_done, NULL);
+	} else {
+		g_target_state = NVMF_FINI_SUBSYSTEM;
+	}
+}
+
+static void
+nvmf_create_nvmf_tgt(void)
+{
+	struct spdk_nvmf_subsystem *subsystem;
+	struct spdk_nvmf_target_opts tgt_opts;
+
+	tgt_opts.max_subsystems = g_nvmf_tgt.max_subsystems;
+	snprintf(tgt_opts.name, sizeof(tgt_opts.name), "%s", "nvmf_example");
```

Comment:
Nit: `nvmf_example` -> real name

Reply:
Regarding license: there are couple of other sources in src/common with bsd license, e.g. sctp_crc32.c or crc32c_intel_baseline.c, so I think this should not be the issue, although IANAL :)

Reply:
... but you are modifying the source. I believe the other cases are just straight copies(?) and they are called out in `COPYING`

Reply:
Seems BSD is compatible with GPL, so I should be in my right to release modifications under GPL _but_ keeping original BSD license text and authors in the COPYING file.

Reply:
Like I said, I would want someone w/ more legal knowledge to make that determination.